File: implementations/kotlin-springboot/examples/src/main/kotlin/com/example/accountservice/payment/application/event/PaymentCompletedEventHandler.kt
package com.example.accountservice.payment.application.event

import com.example.accountservice.outbox.OutboxWriter
import com.example.accountservice.payment.application.integrationevent.PaymentCompletedIntegrationEventV1
import com.example.accountservice.payment.domain.PaymentCompletedEvent
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component

/**
 * The Application EventHandler that receives the internal Domain Event (PaymentCompletedEvent),
 * converts it into the Integration Event for external BCs (payment.completed.v1), and loads it into
 * the Outbox. Account BC subscribes to this Integration Event and performs the actual deduction
 * (withdraw).
 */
@Component
class PaymentCompletedEventHandler(
    private val outboxWriter: OutboxWriter,
) {
    private val logger = LoggerFactory.getLogger(PaymentCompletedEventHandler::class.java)

    fun handle(event: PaymentCompletedEvent) {
        logger
            .atInfo()
            .addKeyValue("payment_id", event.paymentId)
            .addKeyValue("account_id", event.accountId)
            .log("Payment completed")

        outboxWriter.saveAll(
            listOf(
                PaymentCompletedIntegrationEventV1(
                    paymentId = event.paymentId,
                    accountId = event.accountId,
                    amount = event.amount,
                    completedAt = event.completedAt.toString(),
                ),
            ),
        )
    }
}

File: implementations/fastapi/examples/src/payment/application/event/payment_completed_event_handler.py
from __future__ import annotations

import logging

from ....outbox.outbox_writer import OutboxWriter
from ..integration_event.payment_completed_integration_event import PaymentCompletedIntegrationEventV1

logger = logging.getLogger(__name__)


class PaymentCompletedEventHandler:
    """An Application EventHandler that receives the internal Domain Event
    (PaymentCompleted), converts it into an Integration Event for external BCs
    (payment.completed.v1), and loads it into the Outbox. The Account BC subscribes to this
    Integration Event and performs the actual debit (withdraw).
    """

    def __init__(self, outbox_writer: OutboxWriter) -> None:
        self._outbox_writer = outbox_writer

    async def handle(self, payload: dict) -> None:
        logger.info(
            "Payment completed: payment_id=%s account_id=%s amount=%s",
            payload["payment_id"],
            payload["account_id"],
            payload["amount"],
        )
        await self._outbox_writer.save_all(
            [
                PaymentCompletedIntegrationEventV1(
                    payment_id=payload["payment_id"],
                    account_id=payload["account_id"],
                    amount=payload["amount"],
                    completed_at=payload["completed_at"],
                )
            ]
        )
